### DISCUSSION TITLE
v2: add cmds for toml-test integration

### DISCUSSION THREAD

**Discussion by moorereason** [pelletier/go-toml#599] [Ideas]
It would be useful to pass the tests in the [toml-test](https://github.com/BurntSushi/toml-test) project.  We need a cmds similar to [toml-test-decoder](https://github.com/BurntSushi/toml/tree/master/cmd/toml-test-decoder) and [toml-test-encoder](https://github.com/BurntSushi/toml/tree/master/cmd/toml-test-encoder).

I hacked up a partial implementation using some of the BurntSushi json tagging code.  It works with the exception of datetime json tagging.  With my tests, v2 decoder results:  **237 passed, 47 failed** (but 3 of the failures are related to json tagging issues in the integration).

**Comment by pelletier** [comment]
We used to have a binary to be compatible with with toml-test. However I replaced it with a program that generates Go test files for those: https://github.com/pelletier/go-toml/pull/264. I still like its approach for portable testing.

Maybe we should re-gen the tests periodically so that we can catch new stuff?

**Reply by moorereason** [reply]
I'd like to eventually have both (the toml-test binary and a test-gen tool).  For now I think updating the tests-gen tool is a better step to getting us passing all the tests and for CI.

The toml-test framework has changed such that the gen tool doesn't currently work.  I'll see if I can get it updated for v2. 👍

**Reply by pelletier** [reply]
I'm open to it. Can you elaborate on why you'd like both the binary and the test gen?

**Reply by moorereason** [reply]
I'm wanting to use the binary API to compare various decoders' handling of fuzzed input.  Should give more corner cases not in the existing toml-test suite. Already found one bug in BurntSushi/toml.

**Reply by pelletier** [reply]
That's cool! Sounds like we can have both: binaries to test encoding/decoding and generating tests (to avoid having BurntSushi/toml as a dependency or resorting to a script).

**Reply by moorereason** [reply]
As a follow-up to this discussion, I've pushed up my proof-of-concept fuzzing setup:  [tomlfoolery](https://github.com/moorereason/tomlfoolery).  The only two implementations I'm aware of for testing the framework are [go-toml](https://github.com/pelletier/go-toml/tree/v2/cmd/gotoml-test-decoder) and [BurntSushi](https://github.com/BurntSushi/toml/tree/master/cmd/toml-test-decoder).

There are bugs in each package that the fuzzer keeps finding, so it's tricky to keep fuzzing once a failed case is found.  I'm still learning how to use fuzzing, so I'm sure there's more that could be done to make it more useful.

**Reply by pelletier** [reply]
Neat! I'm happy to prioritize working on defects found by fuzzing if it helps.

**Reply by moorereason** [reply]
Most of what is found today is covered by the toml-test suite. Once those all pass, it will get more interesting.
